### DISCUSSION TITLE
How to deploy spack installed software on cvmfs system?

### DISCUSSION THREAD

**Discussion by neutralino-ai** [spack/spack#48048] [Q&A] [answered]
It would be very useful to combine `spack` with `cvmfs`, yet there is little documentation about it. Here I share what I found:

### What is exactly the requests?
- As the software administrator, I need to release the external dependence and offline packages on `/cvmfs/myexp.myuni.edu`.
- It's required that the number of files deployed to this folder to be limited, due to inode restrictions on our computing cluster.
- Thus it would be useful that **only the installed binaries and necessary setup scripts are rsynced to the cvmfs folder, and other files from spack are not copied at all, to save inode.**
- **For the setup scripts, it can be a simple `setup.sh`, or files that can be used through `module use`.**
- Also, for end users, it would be better that they do not need to activate the spack environment, so it would be nice that it's possible that `spack` is not copied to `cvmfs`. End users should only need to `source setup.sh` or `module load xxx` to keep their life easy.

### Solutions (wild guess)
1. `spack view copy`
- in this method, we use spack view to copy installed packages to a folder. The rest files can be deleted, to save inodes. User is responsible to creating bash scripts to setup all environmental variables like `PATH`, `LD_LIBRARY_PATH` etc. 
- I believe automatic generation of such setup scripts is not implemented yet, check #11822 
2. `spack module`
- in this method, we use `spack/opt` folder and `module` folder. The rest files can be deleted, to save inodes. User use `module load` to load packages.
3. `spack buildcache install`
- in this method, we use `spack buildcache push` to push some compressed files to some folder and delete everything else. Then copy this files to another machine. On the other machine, install spack first, then unpack files in this folder with `spack buildcache install`.

I'm not sure if my guess are right / implemented. Here are some literature review:

### Official documents
- related content: [spack-tutorial: binary cache tutorial](https://spack-tutorial.readthedocs.io/en/latest/tutorial_binary_cache.html#relocation). In the tutorial it's mentioned that to release the software, the path inside binary need to be patched, and this is automatic in `spack buildcache install`. I think `spack view copy` also do that, but I need to check the code. At the same time, `spack config add config:install_tree:padded_length:256` should be done.
- related content: [spack: binary_cache-relocation](https://spack.readthedocs.io/en/latest/binary_caches.html#relocation)

### Documents from CERN etc.
- [Using CVMFS with Spack at the FCC study](https://indico.cern.ch/event/757415/contributions/3421522/attachments/1854591/3046155/FCCSW-Spack-CVMFS_Workshop_2019.pdf)
- [eic-spack-cvmfs github repo](https://github.com/eic/eic-spack-cvmfs)
- [key4hep github repo](https://key4hep.github.io/key4hep-doc/spack-build-instructions-for-librarians/spack-setup.html)
- [PR for `spack view copy`](#24832 )
I believe key4hep has already combined cvmfs and spack, without copying whole spack to cvmfs.

No idea whom to @, but let me try: Maybe internals knows whom to @. @psakievich @vsoch 

Many thanks if you could push spack to be used by more experiments.

**Comment by jmcarcell** [comment]
From the Key4hep side what we do is to build in a node in folders with the same name and location as the ones from CVMFS, like `/cvmfs/sw.hsf.org/` and then copy the build to the publisher node with rsync. If you don't want the spack files you could just exclude `.spack` but I feel that depending on which packages you want to install `.spack` is a minority of the files that are deployed. 
For setup we have a `setup.sh` script that sets up all the paths to the packages; most of our users don't use Spack or care / have to know that Spack is being used to build the packages. The scripts are located at:
```
/cvmfs/sw-nightlies.hsf.org/key4hep/setup.sh
```
or
```
/cvmfs/sw.hsf.org/key4hep/setup.sh
```
and they are created by https://github.com/key4hep/key4hep-spack/blob/main/packages/key4hep-stack/common.py#L145

**Comment by psakievich** [comment, marked as answer]
I'm not familiar with cvfms, but I will tell you some tips I've done for deployment that are similar to the goals you've outlined. 
It sounds like the ultimate goal is deploy a stack of binaries and possibly modules from spack, but without needing spack's files, or users to interface with spack going forward? Overall your wild guesses seem pretty on the money to me, but you can combine them to give yourself and your users better interfaces.

One way I have tackled this problem in the past is to generate an environment specifically to deploy that can be deleted later. It's basically a combination of options 2 and 3.

``` yaml
spack:
  config:
     # install the binaries at a location of your choosing with a simpler projection
     install_tree: 
        root: /path/where/you/want/binaries/to/live
        projections:
           all: "{compiler.name}-{compiler.version}/{name}-{version}"
   modules:
       # do the same for modules
       enable: [tcl]
       default:
          roots:
             tcl: /path/where/modules/should/live
          tcl:
            projections:
               all: "{name}-{version}-{compiler.name}-{compiler.version}"
```
With this environment you can add the specs you want, install prebuilt assets from a binary cache and then delete the environment, and really all of spack. Note for these modules you don't need spack at all. They are standard tcl or lmod modules. `spack module` is simply a way to generate, refresh, delete the modules.

So the steps to do this would be:

1. Build out your stack and push to a binary cache (nice if you plan to redeploy/change configurations etc later)
2. Create an environment that configures modules and binaries at a path where you want them deployed (via install tree or view)
3. Add the specs to the environment you want from the cache for this deployment
4. `spack install --cache-only` your environment to install your desired deployment assets
5. `spack module [tcl/lmod] refresh` to generate the module files
6. Delete the environment and/or spack instance after you confirm you are satisfied with the deployment

Another useful feature is the `use_view` described in the later half of the docs [here](https://spack.readthedocs.io/en/latest/module_file_support.html#customize-environment-modifications). You can use this rather than the install_tree if you want.  I'm not currently sure if a copy view can be used with an environment. Some work has been done to make view's a proper configuration, but I'm not sure if the copy/hardlink options have been tested/work with environments. I believe @tldahlgren or @scheibelp worked on this? 

Overall a copy view is a fine way to go.  You could do all of this outside of an environment by just configuring these settings inside some scope in your configuration hierarchy. Environments for this purpose are just a nice way to encapsulate all the changes in one file/place for your deployment methodology of choice.

I think the key take away is that if you want to deploy with minimal files, spack has the mechanisms for you to configure via `view` or `install_tree` such that there is no need for the spack tooling to live on with the deployed assets. Most of the spack literature and endorsed way to deploy for others to avoid spack internals is via view or module.  In this case I would argue a copy view or custom `install_tree` are essentially the same thing.
